[DataGrid] Provide alternate way to refresh when working with remote data

EvandroMoSou opened the pull request:
<!---
Thanks for filing a pull request 😄 ! Before you submit, please read the following:

Search open/closed issues before submitting. Someone may have pushed the same thing before!

Provide a summary of your changes in the title field above.
-->

# Pull Request

## 📖 Description

Using ItemsProvider and Virtualize like the sample in [DataGrid Remote Data](https://www.fluentui-blazor.net/datagrid-remote-data) the API is called 6 or 7 times. Although it has a CancellationToken in the request this is not enough to mitigate the API's resource consumption.

You can test using the code below on the screen, or seeing the calls to the api in the output:

    @using Microsoft.FluentUI.AspNetCore.Components
    
    @inject HttpClient Http
    @inject NavigationManager NavManager
    
    <div style="height: 434px; overflow:auto;" tabindex="-1">
        <FluentDataGrid
                        ItemsProvider="foodRecallProvider"
                        OnRowDoubleClick="@(()=>DemoLogger.WriteLine("Row double clicked!"))"
                        Virtualize="true"
                        ItemSize="46"
                        GenerateHeader="GenerateHeaderOption.Sticky"
                        TGridItem="FoodRecall" >
            <PropertyColumn Title="ID" Property="@(c => c!.Event_Id)" />
            <PropertyColumn Property="@(c => c!.State)" Style="color: #af5f00 ;" />
            <PropertyColumn Property="@(c => c!.City)" />
            <PropertyColumn Title="Company" Property="@(c => c!.Recalling_Firm)" Tooltip="true"/>
            <PropertyColumn Property="@(c => c!.Status)" />
            <TemplateColumn Title="Actions" Align="@Align.End">
                <FluentButton aria-label="Edit item" IconEnd="@(new Icons.Regular.Size16.Edit())" OnClick="@(() => DemoLogger.WriteLine("Edit clicked"))" />
                <FluentButton aria-label="Delete item" IconEnd="@(new Icons.Regular.Size16.Delete())" OnClick="@(() => DemoLogger.WriteLine("Delete clicked"))" />
            </TemplateColumn>
        </FluentDataGrid>
    </div>
    
    <p>Total: <strong>@numResults results found</strong></p>
    <p>Call Times: <strong>@qtdCall</strong></p>
    
    
    @code {
        GridItemsProvider<FoodRecall> foodRecallProvider = default!;
        int? numResults;
	    int qtdCall;
    
        protected override async Task OnInitializedAsync()
        {
            // Define the GridRowsDataProvider. Its job is to convert QuickGrid's GridRowsDataProviderRequest into a query against
            // an arbitrary data soure. In this example, we need to translate query parameters into the particular URL format
            // supported by the external JSON API. It's only possible to perform whatever sorting/filtering/etc is supported
            // by the external API.
            foodRecallProvider = async req =>
            {
                qtdCall++;
                await InvokeAsync(StateHasChanged);
    
                var url = NavManager.GetUriWithQueryParameters("https://api.fda.gov/food/enforcement.json", new Dictionary<string, object?>
                        {
                    { "skip", req.StartIndex },
                    { "limit", req.Count },
                        });
    
                var response = await Http.GetFromJsonAsync<FoodRecallQueryResult>(url, req.CancellationToken);
    
                // Simulate a slow data retrieval process
                if (req.Count is null)
                {
                    await Task.Delay(2500);
                }
                return GridItemsProviderResult.From(
                    items: response!.Results,
                    totalItemCount: response!.Meta.Results.Total);
            };
    
            // Display the number of results just for information. This is completely separate from the grid.
            numResults = (await Http.GetFromJsonAsync<FoodRecallQueryResult>("https://api.fda.gov/food/enforcement.json"))!.Meta.Results.Total;
        }
    }

After several failed attempts due to multiples refresh by QuickGrid and loss of sequence due to cancellation tokens. I decided to deal with the problem with a different approch.

Normally the API itself takes care of filtering, sorting and pagination, so a new type of grid update was created the 'RefreshItems'.

Basically you can define a method that will call the API.
This method is called once and only once when there is a change in the request (initial, pagination, ordering).
You must update the 'Items' and the 'SetTotalItemCountAsync'.

Along with the necessary commit, the page 'DataGridRemoteData.razor' was updated so you can see how it works.
The display of records, pagination and sorting are working.

Additionally, the option to force the datagrid refresh was added, so that if there is an API filter on the screen that has been modified, you can treat this refresh as a change in the requset.
And a new property in the column 'SortName' to facilitate the assembly of the request for the api.

### 🎫 Issues

[#1066](https://github.com/microsoft/fluentui-blazor/issues/1066)

## 👩‍💻 Reviewer Notes

Start the application and go to "/datagrid-remote-data" you can test here.

## 📑 Test Plan

I did it in a way that the old datagrid functions will not be affected if you have not defined the 'RefreshItems'.
Refresh force is optional and is set to false, so nothing changes for old things either.
And I believe that the new property in the column does not affect anything

## ✅ Checklist

### General

<!--- Review the list and put an x in the boxes that apply. -->

- [ ] I have added tests for my changes.
- [X ] I have tested my changes.
- [ ] I have updated the project documentation to reflect my changes.
- [ ] I have read the [CONTRIBUTING](https://github.com/microsoft/fluentui-blazor/blob/master/docs/contributing.md) documentation and followed the [standards](https://www.fast.design/docs/community/code-of-conduct/#our-standards) for this project.

### Component-specific

<!--- Review the list and put an x in the boxes that apply. -->
<!--- Remove this section if not applicable. -->

- [ ] I have added a new component
- [ ] I have added [Unit Tests](https://github.com/Microsoft/fluentui-blazor/blob/master/unit-tests.md) for my new component
- [X] I have modified an existing component
- [ ] I have validated the [Unit Tests](https://github.com/Microsoft/fluentui-blazor/blob/master/unit-tests.md) for an existing component

## ⏭ Next Steps

Check the interaction of the grid's internal filtering with the new method

EvandroMoSou commented:
I agree with the CLA

EvandroMoSou commented:
@microsoft-github-policy-service agree

vnbaaij reviewed (changes requested):
Thanks for your contribution! Sorry it took a while to review.

There is basically only 1 change I want to see: Create a new page for the altered DataGridRemoteData sample and leave the 'old' one as-is. Now that we have separate grid sample pages it is easy to have (and good) to have more. People can compare the code of both samples to find the differences. And maybe you can also explain what the differences are  in the sample page itself.

vnbaaij commented on examples/Demo/Shared/SampleData/FoodRecall.cs:
```diff
@@ -10,6 +10,7 @@ public class FoodRecall
     public string City { get; set; }
     public string State { get; set; }
     public string Recalling_Firm { get; set; }
+    public string Termination_Date { get; set; }
```
What is the purpose of adding this date? If you do not need it for running your sample then please leave as-is

EvandroMoSou replied:
I use this field to demonstrate the ordering in my example.

vnbaaij commented on src/Core/Components/DataGrid/Columns/ColumnBase.razor.cs:
```diff
@@ -93,6 +93,9 @@ public abstract partial class ColumnBase<TGridItem>
     [Parameter]
     public RenderFragment? ColumnOptions { get; set; }
 
```
You need to add valid XML comment for this (it shows up in the demo site in the API documentation)

EvandroMoSou replied:
Added

vnbaaij commented on src/Core/Components/DataGrid/FluentDataGrid.razor.cs:
```diff
@@ -53,6 +53,9 @@ public partial class FluentDataGrid<TGridItem> : FluentComponentBase, IHandleEve
     [Parameter]
     public IQueryable<TGridItem>? Items { get; set; }
 
+    [Parameter]
```
Also add valid summary comment here

EvandroMoSou replied:
Added

vnbaaij commented on src/Core/Components/DataGrid/FluentDataGrid.razor.cs:
```diff
@@ -691,24 +698,39 @@ private async Task RefreshDataCoreAsync()
             await _virtualizeComponent.RefreshDataAsync();
             _pendingDataLoadCancellationTokenSource = null;
+
+            StateHasChanged();
```
Are you absolutely sure this `StateHasChanged` is needed here?

EvandroMoSou commented:
I reverted the old page, added the new example and fulfilled the requests mentioned.

If you have any other points to adjust, I am at your disposal.

vnbaaij commented:
Thanks for a great contribution!

Just a heads-up, your `SortName` addition will get changed through the word done in #3460.
